File: client/src/main/scala/stasis/client/encryption/secrets/UserHashedEncryptionPassword.scala
package stasis.client.encryption.secrets

import at.favre.lib.hkdf.HKDF
import org.apache.pekko.util.ByteString

import stasis.shared.model.users.User
import stasis.shared.secrets.SecretsConfig

final case class UserHashedEncryptionPassword(
  user: User.Id,
  private val hashedPassword: ByteString
)(implicit target: SecretsConfig)
    extends Secret {
  def toLocalEncryptionSecret: UserLocalEncryptionSecret = {
    val salt = user.toBytes
    val keyInfo = ByteString(s"${user.toString}-encryption-key")
    val ivInfo = ByteString(s"${user.toString}-encryption-iv")

    val hkdf = HKDF.fromHmacSha512()

    val prk = hkdf.extract(salt, hashedPassword.toArray)

    val key = hkdf.expand(prk, keyInfo.toArray, target.encryption.deviceSecret.keySize)
    val iv = hkdf.expand(prk, ivInfo.toArray, target.encryption.deviceSecret.ivSize)

    UserLocalEncryptionSecret(user = user, key = ByteString(key), iv = ByteString(iv))
  }

  def toKeyStoreEncryptionSecret: UserKeyStoreEncryptionSecret = {
    val salt = user.toBytes
    val keyInfo = ByteString(s"${user.toString}-key-store-encryption-key")
    val ivInfo = ByteString(s"${user.toString}-key-store-encryption-iv")

    val hkdf = HKDF.fromHmacSha512()

    val prk = hkdf.extract(salt, hashedPassword.toArray)

    val key = hkdf.expand(prk, keyInfo.toArray, target.encryption.deviceSecret.keySize)
    val iv = hkdf.expand(prk, ivInfo.toArray, target.encryption.deviceSecret.ivSize)

    UserKeyStoreEncryptionSecret(user = user, key = ByteString(key), iv = ByteString(iv))
  }
}

object UserHashedEncryptionPassword {
  def apply(
    user: User.Id,
    hashedPassword: ByteString
  )(implicit target: SecretsConfig): UserHashedEncryptionPassword =
    new UserHashedEncryptionPassword(user, hashedPassword)
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/encryption/secrets/UserHashedEncryptionPassword.kt
package stasis.client_android.lib.encryption.secrets

import at.favre.lib.hkdf.HKDF
import okio.ByteString
import okio.ByteString.Companion.encodeUtf8
import okio.ByteString.Companion.toByteString
import stasis.client_android.lib.model.server.users.UserId

data class UserHashedEncryptionPassword(
    val user: UserId,
    private val hashedPassword: ByteString,
    val target: Config
) : Secret() {
    fun toLocalEncryptionSecret(): UserLocalEncryptionSecret {
        val salt = user.toBytes()
        val keyInfo = "${user}-encryption-key".encodeUtf8()
        val ivInfo = "${user}-encryption-iv".encodeUtf8()

        val hkdf = HKDF.fromHmacSha512()

        val prk = hkdf.extract(salt, hashedPassword.toByteArray())

        val key = hkdf.expand(prk, keyInfo.toByteArray(), target.encryption.deviceSecret.keySize)
        val iv = hkdf.expand(prk, ivInfo.toByteArray(), target.encryption.deviceSecret.ivSize)

        return UserLocalEncryptionSecret(
            user = user,
            key = key.toByteString(),
            iv = iv.toByteString(),
            target = target
        )
    }

    fun toKeyStoreEncryptionSecret(): UserKeyStoreEncryptionSecret {
        val salt = user.toBytes()
        val keyInfo = "${user}-key-store-encryption-key".encodeUtf8()
        val ivInfo = "${user}-key-store-encryption-iv".encodeUtf8()

        val hkdf = HKDF.fromHmacSha512()

        val prk = hkdf.extract(salt, hashedPassword.toByteArray())

        val key = hkdf.expand(prk, keyInfo.toByteArray(), target.encryption.deviceSecret.keySize)
        val iv = hkdf.expand(prk, ivInfo.toByteArray(), target.encryption.deviceSecret.ivSize)

        return UserKeyStoreEncryptionSecret(
            user = user,
            key = key.toByteString(),
            iv = iv.toByteString(),
            target = target
        )
    }
}
